text/0000-convert-id.md

```diff
@@ -0,0 +1,201 @@
+- Feature Name: convert_id
+- Start Date: 2018-01-19
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Adds an identity function `pub fn id<T>(x: T) -> T { x }` as `core::convert::id`.
```

Comment:
I feel it should be called `identity`, id is really short and ambiguous.

Reply:
`id` has pecedent in other languages like Haskell. I could go either way, but it being short and known as the identity function elsewhere is nice to have.

Reply:
Identity is a better name in the prelude, if we keep that possibility open. I think we go for not overly abbreviated names of functions in Rust

Reply:
It's also called identity in Java, which is probably more used than Haskell. Though that function may be used less in Java.

Familiarity arguments work if there's uniformity; there isn't, and it feels like plenty of folks won't have ever consciously noticed/used it, so I'd prefer to optimize for folks who don't know what it is. `identity` is clear and also should be obvious to those used to `id`

Reply:
@Manishearth I've provided this motivation for the naming:
> Naming the function `identity` instead of `id` is a possibility. However, to make the `id` function more appetizing than using a `|x| x`, it is preferrable for the `identity` function to have a shorter but still clear name.

Why is `id` ambiguous? Are you referring to a potential ambiguity with "identifier"?
And why is this potential ambiguity an overriding concern to the one I've listed?
I think `id` is clear from context - If there are other functions around in the context that look like `T -> T`, you know it's not going to be "identifier"...

Yes - I suspect it is used way less in Java than in Haskell. I prefer to optimize for folks who are most likely to use the operation, which are probably on balance more functional programmers.

Reply:
I do not find length to be that pressing a concern when it comes to ergonomics.

Yes, id sounds like identifier. Not everyone will immediately make the connection with the identity function; and without the math background many won't know what that _is_. They may be able to guess, but not if you call it `id`.

It's also very short and putting more short things in the prelude is not something I really like (but I can't articulate why :/ )

Reply:
@bluss As a general rule I think that we should have descriptive and not overly abbreviated names - but every rule has its exception and I think making one in this case is motivated (not by familiarity).

Reply:
> I do not find length to be that pressing a concern when it comes to ergonomics.

Just to clarify: it is the relative length I am concerned about, not the absolute. I'd like to encourage people to use `id` instead of `|x| x`, but that might be harder with `identity`... Does that make sense to you?

> [..] more short things [..]

Out of curiosity - since you say **more** short things - which ones are you referring to?

> [..] (but I can't articulate why :/ ) [..]
But if you can, that will make your concern more actionable for me =)

Reply:
My point still stands. `identity` is more ergonomic than the closure even if it is longer, by virtue of being just an identifier. It also can be used in more contexts because it has the same type as other functions (almost).

I do not think that I'd being shorter than the closure has any benefit. We are rarely trying to optimize for less typing. [See also](https://twitter.com/withoutboats/status/950582458797273088)

We already have short functions like drop in the prelude. drop is fine and well known, but adding more feels iffy to me. I feel this is going to clash with locals quite often. Clashes in the prelude are fine, but make things murkier because someone reading the code would expect it to be a local.

(Actually, for that matter there's a general argument to be made against adding functions to the prelude -- there are _very few_ and it's going to be surprising for folks to see a different function that doesn't seem to be imported anywhere. Unless it becomes ubiquitous. Which I suspect it won't.)

Reply:
I guess I'll change it to `identity` then by popular support. What tipped me over was this note by @nox:
> 88,618 lines with 'id' in Servo. 

I think it is necessary and essential that this be in the prelude however and there's certainly precedent for it.

Reply:
How embarrassing.

For the record, I forgot to filter the grep command to only `*.rs` files and thus it also included our local clone of `web-platform-tests`, a massive repository of Web tests running in browsers. There are actually only 1,533 lines with the word `id` in Servo.

That's still a huge number to me though, but nowhere near the one I mistakenly reported at first.

Reply:
And the function shall henceforth be known as `identity`.